paddle/fluid/operators/shape_op.cu

```diff
@@ -12,12 +12,58 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. */
 
+#include <cuda.h>
+#include <vector>
 #include "paddle/fluid/operators/shape_op.h"
 
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+using LoDTensor = framework::LoDTensor;
+using SelectedRows = framework::SelectedRows;
+
+template <typename T>
+class ShapeCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* in_var = ctx.InputVar("Input");
+    framework::DDim in_dims;
```

Comment:
Not important here but you should notice when you write code elsewhere. C++ is a complex language so we often have to distinguish the reference assign and value copy very clearly to make code faster.

framework::DDim overrides the operator= as a copy. So here your code initializes a DDim and then copies the DDim, which slows down this shape_op. A faster way will be using a const pointer or const reference. For example:

```
const framework::DDim &in_dims = in_var->IsType<SelectedRows>() ? in_var->Get<SelectedRows>().value().dims() : in_var->Get<LoDTensor>().dims();
```

It is not important here because shape_op usually is not the speed bottleneck. So you can choose to change or not change here but I think it is good for you to know it (maybe you already know it then it is just a reminder) so I wrote this long comment :-)